### ISSUE TITLE
Support TypeScript out of the box

### ISSUE THREAD

**Issue by sapegin** [styleguidist/react-styleguidist#750]
- [x] Upgrade react-docgen to 5.0.0 beta (#1354)
- [x] Tweak default `components` and `ignore` patterns to include `.ts` and `.tsx`.
- [x] Tweak [default extensions](https://github.com/styleguidist/react-styleguidist/blob/0ef7ff38a550b7b049bd0b82d87bd2ec97af3fce/src/scripts/make-webpack-config.js#L45) in webpack config to include `.ts` and `.tsx`. 
- [x] Update docs maybe (search for `typescript` and see if the information is still relevant).

**Comment by stereobooster** [comment]
I do not want to open an separate issue, because I'm not sure if this an issue of `styleguidist` or my setup. I bootstrapped project with https://github.com/wmonk/create-react-app-typescript

My config is:

```js
const path = require("path");
module.exports = {
  styleguideComponents: {
    Wrapper: path.join(__dirname, "src/Wrapper")
  },
  skipComponentsWithoutExample: true,
  propsParser: require("react-docgen-typescript").withCustomConfig(
    "./tsconfig.json"
  ).parse
};
```

I get the following error

```
InvalidCharacterError: Failed to execute 'createElement' on 'Document': The tag name provided ('static/media/Navbar.b192c68a.tsx
```

**Comment by nwinger** [comment]
Experiencing the same issue as @stereobooster running `create-react-app my-app --scripts-version=react-scripts-ts`.

My config looks like this:
```javascript
const path = require('path');

module.exports = {
  components: 'src/components/**/*.{ts,tsx}',
  propsParser: require('react-docgen-typescript').parse,
  webpackConfig: require('react-scripts-ts/config/webpack.config.dev.js'),
  styleguideComponents: {
    StyleGuideRenderer: path.join(
      __dirname,
      'lib/styleguide/StyleGuideRenderer'
    )
  }
};
```

**Comment by nwinger** [comment]
Update:
Had a colleague look at this, and he suggested we looked into the webpack.config.dev.js from react-scripts-ts. Sure enough, I'd placed my "lib" folder where the docs suggested, but I had to move it inside of source as the loader is looking in that directory for ts and tsx files.

**Comment by okonet** [comment]
@sapegin I might need this and could work on this. I'm not sure about the second step, though. Any pointers?

**Comment by sapegin** [comment]
I think your approach with extension checking is the right one!

**Comment by sapegin** [comment]
> Make react-docgen-typescript read users’s tsconfig.json by default.

I'm not sure about that after looking at the code ;-/

**Comment by okonet** [comment]
Yep, we should resolve the path carefully.

**Comment by tsiq-jeremy** [comment]
I was doing a little bit of research on this and it looks like [react-docgen is trying to support typescript](https://github.com/reactjs/react-docgen/issues/325), but it seems to be in early stages. If they did support TS there would not be a need for react-docgen-typescript anymore. 

> Automatically enable react-docgen-typescript for TypeScript projects (check typescript in devDependencies?

Seems doable, just update props-loader.js to check if TS is supported then use react-docgen-typescript if it is.

**Comment by s-h-a-d-o-w** [comment]
In case you are still considering this, I would recommend you run react-docgen-typescript against some slightly complex cases such as a formerly Flow-typed FC that destructures props, assigns default values to them and uses typed objects for some of the props - some of which are optional.

Unfortunately, none of that works well, as I have discovered today, after migrating our UI component library at work from Flow to TS. Default values are missing, the structure of typed objects is not passed to styleguidist and optional props are described as `propType | undefined`.
And so we are currently struggling with whether we can fix this in a reasonable amount of time or have to build our API docs some other way. But we've enjoyed using styleguidist with Flow-typed components very much, so we're hoping for the former.

But that's why I'd recommend waiting for that react-docgen TS support that @tsiq-jeremy mentioned.

**Comment by s-h-a-d-o-w** [comment]
The beautiful thing is... possibly due to babel magic (gotta love it), it looks like people who need TS support right now can actually already have it. I've played around with the `Button` component in the "basic" example:
https://github.com/s-h-a-d-o-w/react-styleguidist/tree/basic_TS_example

Looks just wonderful to me. Not sure whether it has any flaws but... I'll definitely try this out on our library at work soon. :wink:

**Comment by devongovett** [comment]
FYI, I'm working on TypeScript support for react-docgen here. Feel free to test it out: https://github.com/reactjs/react-docgen/pull/348

**Comment by J-Kallunki** [comment]
Should this also include support for customizing rsg-components? Got typescript working fine for presented components, but creating custom rsg-components with `styleguideComponents` seems to be a bit harder when using TS.

**Comment by sapegin** [comment]
@J-Kallunki which problems do you have? There shouldn't be any difference ;-/

**Comment by J-Kallunki** [comment]
I'm using Webpack and awesome-typescript-loader.

`ComponentsList.tsx`:
```
import React, { Component } from 'react';
import { default as DefaultComponentsList } from 'react-styleguidist/lib/client/rsg-components/ComponentsList/ComponentsList';
export default class ComponentsList extends Component {
  render() {
    return <DefaultComponentsList {...this.props} />;
  }
}
```

Typings (minimal):
```
declare module 'react-styleguidist/lib/client/rsg-components/ComponentsList/ComponentsList' {
  interface ComponentsListProps {
    [key: string]: any | any[];
  }

  const ComponentsList: React.FunctionComponent<ComponentsListProps>;
  export default ComponentsList;
}
```

Result:
```
./node_modules/react-styleguidist/lib/client/rsg-components/ComponentsList/ComponentsList.js
Module not found: Can't resolve 'rsg-components/ComponentsList/ComponentsListRenderer'
```

So I would need to configure babel just for it and use only non-TS custom components?

**Comment by sapegin** [comment]
I don't think you need to configure Babel, everything in `lib` is already compiled. But you need to configure TypeScript for your `ComponentsList.tsx`.  Or you many need to define typings for `rsg-components/ComponentsList/ComponentsListRenderer` too, I don't know TypeScript enough unfortunately.

**Comment by J-Kallunki** [comment]
> I don't think you need to configure Babel, everything in `lib` is already compiled. But you need to configure TypeScript for your `ComponentsList.tsx`. Or you many need to define typings for `rsg-components/ComponentsList/ComponentsListRenderer` too, I don't know TypeScript enough unfortunately.

I got the earlier example working just by defining to `moduleAliases`:
```
'rsg-components/ComponentsList/ComponentsListRenderer': path.resolve(__dirname, 'node_modules/react-styleguidist/lib/client/rsg-components/ComponentsList/ComponentsListRenderer'),
```

Is it the best way to do it - it is the simplest - or should it be also documented how to use custom srg components with TS?

**Comment by sapegin** [comment]
Have you tried this?

> Or you many need to define typings for rsg-components/ComponentsList/ComponentsListRenderer too

**Comment by J-Kallunki** [comment]
> Have you tried this?
> 
> > Or you many need to define typings for rsg-components/ComponentsList/ComponentsListRenderer too

Yeah tried that with the same way as it works for the `ComponentsList` but not helping for that.

It is clearly saying that cannot find `ComponentsListRenderer` when `ComponentsList`is called from the customwrapper `ComponentsList`-component:
```
Module not found: Can't resolve 'rsg-components/ComponentsList/ComponentsListRenderer'
```

**Comment by sapegin** [comment]
@J-Kallunki I may have fixed that in 9e29c10dfa97cf56b5aa55f0483de5c3dad3bb5b, at least I had the same error ;-)

**Comment by plumpNation** [comment]
I am not certain this is the same issue, apologies if not. I'm using **create-react-app**, I made a tsx `styleguideComponents` Wrapper, and setup `propsParser: require('react-docgen-typescript').parse`.

```
"react-docgen-typescript": "^1.12.4",
"react-styleguidist": "^9.1.6"
```

I created the smallest reproduction of this that I could and made a repo here.

https://github.com/plumpNation/cra-styleguidist-typescript

Thanks for the excellent tool, I will just switch the wrapper to jsx until I find out how to support tsx styleguideComponents.
